gradle-baseline-java/src/main/groovy/com/palantir/baseline/tasks/CheckJUnitDependencies.java

```diff
@@ -0,0 +1,171 @@
+                            + "add the following:\n\n"
+                            + "    " + implementation + " 'org.junit.jupiter:junit-jupiter'\n");
+        }
+
+        // When doing an incremental migration to JUnit5, a project may have some JUnit4 and some JUnit5 tests at the
+        // same time. It's crucial that they have the vintage engine set up correctly, otherwise tests may silently
+        // not run!
+        if (sourceSetMentionsJUnit4(ss)) {
+            if (BaselineTesting.useJUnitPlatformEnabled(task)) { // people might manually enable this
+                String testRuntimeOnly = ss.getRuntimeConfigurationName() + "Only";
+                Preconditions.checkState(
+                        junitJupiterIsPresent,
+                        "Tests may be silently not running! Some tests still use JUnit4, but Gradle has "
+                                + "been set to use JUnit Platform. "
+                                + "To ensure your old JUnit4 tests still run, please add the following:\n\n"
+                                + "    " + testRuntimeOnly + " 'org.junit.jupiter:junit-jupiter'\n\n"
+                                + "Otherwise they will silently not run.");
+
+                boolean vintageEngineExists = hasDep(
+                        ss.getRuntimeClasspathConfigurationName(), CheckJUnitDependencies::isVintageEngine);
+                Preconditions.checkState(
+                        vintageEngineExists,
+                        "Tests may be silently not running! Some tests still use JUnit4, but Gradle has "
+                                + "been set to use JUnit Platform. "
+                                + "To ensure your old JUnit4 tests still run, please add the following:\n\n"
+                                + "    " + testRuntimeOnly + " 'org.junit.vintage:junit-vintage-engine'\n\n"
+                                + "Otherwise they will silently not run.");
+            } else {
+                Preconditions.checkState(
+                        !junitJupiterIsPresent,
+                        "Tests may be silently not running! Please remove "
+                                + "'org.junit.jupiter:junit-jupiter' dependency "
+                                + "because tests use JUnit4 and useJUnitPlatform() is not enabled.");
+            }
+        } else {
+            String compileClasspath = ss.getCompileClasspathConfigurationName();
+            boolean compilingAgainstOldJunit = hasDep(compileClasspath, CheckJUnitDependencies::isJunit4);
+            Preconditions.checkState(
+                    !compilingAgainstOldJunit,
+                    "Extraneous dependency on JUnit4 (no test mentions JUnit4 classes). Please exclude "
+                            + "this from compilation to ensure developers don't accidentally re-introduce it, e.g.\n\n"
+                            + "    configurations." + compileClasspath + ".exclude module: 'junit'\n\n");
```

Comment:
This might ding quite a lot of projects, but I think the consistency is probably worth it.